nixos/xserver: use modern video drivers

TredwellGit opened the pull request:
###### Motivation for this change
cirrus is obsolete: https://www.vintage3d.org/cirrus.php
nv is obsolete: https://www.phoronix.com/scan.php?page=article&item=nvidia_kills_nv&num=1
vesa is obsolete: https://www.phoronix.com/scan.php?page=news_item&px=Nzc3Nw
ati and ati_unfree are superseded by amdgpu and amdgpu-pro: https://wiki.gentoo.org/wiki/ATI_FAQ#Is_my_AMD.2FATI_board_supported.3F
nouveau and fbdev added for better fallback support.

###### Things done
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

TredwellGit commented:
Also can someone merge https://github.com/NixOS/nixpkgs/pull/98974?

buckley310 commented:
Working here.

I applied this change, removed `xserver.videoDrivers` from my config, and rebooted. Amdgpu driver was loaded.
I'm using an AMD 5700XT.

Colecf commented:
Please don't remove vesa from the defaults. I have an nvidia RTX 3080, and the nix-packaged version of the nvidia driver is currently version 440.100, which doesn't support my card. Vesa is the only video driver that works for me.

Of course, it doesn't support the full display resolution, but its much better to at least get into a functioning version of the desktop as opposed to being locked out and forced to roll back to an earlier nixos generation.

TredwellGit commented:
I would prefer if someone just merged and backported https://github.com/NixOS/nixpkgs/pull/98191.

Atemu commented on nixos/modules/services/x11/xserver.nix:
```diff
@@ -248,11 +248,10 @@ in
       videoDrivers = mkOption {
         type = types.listOf types.str;
-        # !!! We'd like "nv" here, but it segfaults the X server.
-        default = [ "radeon" "cirrus" "vesa" "modesetting" ];
```
What's the reason for removing cirrus and vesa here?   

Putting them as a lower priority than modesetting would be appropriate I think.

TredwellGit replied:
Did you read the articles? Both cirrus and vesa are obsolete.

thiagokokada replied:
Any reason we don't add `nouveau` here?

thiagokokada replied:
BTW, having `vesa` as the last option seems interesting at least. In case everything else fails at least the user would have a "working" X11 (and in the linked article provided from Phoronix this is exactly the use case they cite as valid, as a fallback driver).

TredwellGit replied:
Nouveau can't reclock modern GPUs and lacks Vulkan support. See pages [2](https://www.phoronix.com/scan.php?page=article&item=opensource-turing-3d&num=2) and [3](https://www.phoronix.com/scan.php?page=article&item=opensource-turing-3d&num=3) of [this](https://www.phoronix.com/scan.php?page=article&item=opensource-turing-3d) article. I don't see why anyone would purchase an NVIDIA GPU and want to use anything but the `nvidia` video drivers. If you cared about using libre software you would not have purchased NVIDIA in the first place.

thiagokokada replied:
Does `modesetting` support reclock/Vulkan in NVIDIA? If not, I am still in favor of adding them to `default`. The `default` exists to have at least something working for the use, however if the user wants to use proprietary solutions they can overwrite this setting (and this is why it exists in first place).

> If you cared about using libre software you would not have purchased NVIDIA in the first place.

Please, no pointing fingers here. I don't use NVIDIA GPUs in Linux exactly because of their bad behavior, but my employee for example gave me a laptop with NVIDIA GPU. There are probably other users out there in the same situation, and I think offering them a good default is better.

TredwellGit replied:
I am not pointing fingers. I am just saying that if there is a modern NVIDIA GPU, then `nvidia` should be used. The default cannot have `nvidia` because it is incompatible with everything else. Nothing but `nvidia` supports reclocking since NVIDIA requires cryptographically signed firmware images to force usage of `nvidia`.

thiagokokada replied:
But I am not asking for adding `nvidia` (I frankly hate this proprietary driver, there is so many issues and things broken), I am asking why you can't add `nouveau` for a better default user experience?

I mean, AFAIK (I may be wrong), if we have `nouveau` in this list and the user wants to use `nvidia`, they can simple set `xserver.videoDrivers = [ "nvidia" ]`.

And why adding `nouveau` it is because, again AFAIK, it offers basic 2D and 3D acceleration, sufficient for desktop usage. I don't think `modesetting` offers this for NVIDIA users (or that it even works at all), so this is why I think adding `nouveau` would be interesting.

TredwellGit replied:
What I am trying to understand is why someone would want to use `nouveau` at all when `nvidia` is available considering how `nvidia` is 16 to 20 times faster than `nouveau`, supports Vulkan, supports modern OpenGL, supports CUDA, etc.

TredwellGit replied:
Also, the only situation I am aware of where `vesa` would be necessary is when Nixpkgs has an outdated version of `nvidia` and someone is trying to use a GPU that requires a newer version. In which case I would much prefer just to keep `nvidia` updated.

thiagokokada replied:
Let's say I am a person using NixOS for the first time ever. I am probably overwhelmed about all the options and trying to get a system to work. If we put `nouveau` in the default drivers this newbie will not need to worry with drivers until it tries a game and it doesn't work (or work slowly), but at least the desktop will work (20 times slower than official drivers should still be sufficient to get good enough performance in desktop). So after sometime that they feel more confident with NixOS they can work to fix those issues.

If we don't it will maybe not have any desktop at all (exiting with a cryptic X11 error), or maybe it will work but very slowly (both KDE and GNOME expect at least basic 3D accel those days). So the user will get frustrated and decide that NixOS is not for them.

thiagokokada replied:
This is also why I think having `vesa` as the last option is also important. The user will at least get some video (but in this case the experience will be very poor), but it still better than no video at all.

thiagokokada replied:
BTW, if `modesetting` works fine and has basic 2D and 3D accel for NVIDIA GPUs I am fine for not adding `nouveau`. It is just that I am not finding any reference that it works, at all (the only hits I found are about using `modesetting` driver in the Intel GPU for Optimus setup, not for driving a NVIDIA GPU).

TredwellGit replied:
I will add `nouveau` to the defaults, but I feel that better support for enabling `nvidia` when NVIDIA hardware is present should be implemented.

I will submit a pull request to update `xorg.xf86videonouveau` to the latest version and update this pull request later today.

Atemu replied:
Because it's Free, good enough for everything but heavy computation and 10x less of a hassle than nGreedia's blobby crap.

The only use case for the Nvidia driver is when you need nothing but performance and in that case you'd enable it explicitly, which you'd have to do anyways.

TredwellGit replied:
As far as I am aware, `nouveau` is better than `modesetting` for NVIDIA.

TredwellGit replied:
With `default = [ "amdgpu" "nouveau" "modesetting" ];` is there any situation where `vesa` would be needed?

thiagokokada replied:
`vesa` is for the cases when you're using a very obscure GPU with no support at all. It should work with those S3, Cirrus, maybe even some ARM GPUs :thinking: ?

I think it basically works with everything that offers a VESA standard display.

TredwellGit replied:
But does anyone actually use those?

TredwellGit replied:
Also, `modesetting` is supposed to be the modern catchall in that situation.

thiagokokada replied:
Well, like I said it is a fallback, it shouldn't happen but if it does it is still a better user experience than a cryptic X11 error.

I know from some modernish servers (circa ~2010) that still shipped with Matrox GPUs for example.

thiagokokada replied:
> Also, modesetting is supposed to be the modern catchall in that situation.

Well, this is the fallback of the fallback.

thiagokokada replied:
And BTW, like in comment https://github.com/NixOS/nixpkgs/pull/100402#issuecomment-711104497, it likes that `modesetting` failed as a fallback for that case.

Maybe it would work with `nouveau`? Maybe, but this GPU is so new that I can picture it happening at least for a small percentage of the user base on stable at each release.

TredwellGit replied:
Annoyingly, if `vesa` is present it gets priority over the superior `modesetting` which will cause issues with Intel GPUs. I really doubt that this level of compatibility with old obscure GPUs at the expense of modern ones by default is really necessary.

thiagokokada replied:
It this really true? Because I use the old default (`default = [ "radeon" "cirrus" "vesa" "modesetting" ];`) and my driver is correctly set as `vesa`.

Are there any user reports of this issue? If yes, I think it is ok to not add `vesa`, but I found it strange since it works for me.

thiagokokada replied:
BTW, this looks really a case of issue upstream. It doesn't make sense for `vesa` to get priority over `modesetting`, specially if `modesetting` is tried before `vesa`.

TredwellGit replied:
Also, if the GPU does not support kernel mode-setting then most Wayland compositors will not work.

thiagokokada replied:
Huh... Maybe it is my `initrd.kernelModules = [ "i915" ];`. Let me try to remove it and see if it doesn't work.

thiagokokada replied:
> Also, if the GPU does not support kernel mode-setting then most Wayland compositors will not work.

I know, but X11 works and it should be fine for debugging.

I mean, if you have a working X11 you can at least open a graphical browser and look for the solution for your issue. If you have not, you either have to use a terminal browser or another device (that may or may not be available, I was already in situations where my only working device was my notebook/PC).

thiagokokada replied:
> Huh... Maybe it is my initrd.kernelModules = [ "i915" ];. Let me try to remove it and see if it doesn't work.

Nope, I am not using it already. Seems fine to me.

TredwellGit replied:
According to https://wiki.archlinux.org/index.php/Xorg#Driver_installation it is true.

thiagokokada replied:
> According to https://wiki.archlinux.org/index.php/Xorg#Driver_installation it is true.

I think they're referring to they own version of the package, but we are explicitly setting the opposite. Even in how it currently is (`vesa` before `modesetting`), it is working fine, so I think we can keep `vesa` there as the last option).

thiagokokada replied:
Or a probably better hypothesis is that this Wiki is out-dated. Or we can't explain why it works fine now.

TredwellGit replied:
In any case, I am just thinking "who cares?". Who is actually running X11 on a GPU that does not support `modesetting` on a server from 2010-2015 on NixOS? Keep in mind that by putting `vesa` as a default means that it will probably be installed on most graphical systems that use NixOS. Considering that X.Org is in maintenance mode and just exists to support the transition to Wayland which basically requires kernel mode-setting and that this pull request is "use modern video drivers" why are we wasting time, bandwidth, storage, supporting this kind of hardware?

TredwellGit replied:
For example, is there anyone at all running a NixOS system right now where they need `vesa` over `modesetting`.

thiagokokada replied:
Because cases like this may happen: https://github.com/NixOS/nixpkgs/pull/100402#issuecomment-711104497

> Considering that X.Org is in maintenance mode and just exists to support the transition to Wayland 

X.org will keep going for a long, long time.

> why are we wasting time, bandwidth, storage, supporting this kind of hardware?

I am sure that we waste much more bandwidth in other things than a vesa driver:

```
ls -lah /nix/store/jacs1rq2jwl5sxs8hws13rxls31llin6-xf86-video-vesa-2.4.0/lib/xorg/modules/drivers/vesa_drv.* 
-r-xr-xr-x 1 root root 980 dez 31  1969 /nix/store/jacs1rq2jwl5sxs8hws13rxls31llin6-xf86-video-vesa-2.4.0/lib/xorg/modules/drivers/vesa_drv.la
-r-xr-xr-x 1 root root 38K dez 31  1969 /nix/store/jacs1rq2jwl5sxs8hws13rxls31llin6-xf86-video-vesa-2.4.0/lib/xorg/modules/drivers/vesa_drv.so
```

thiagokokada replied:
> For example, is there anyone at all running a NixOS system right now where they need vesa over modesetting.

At least one: https://github.com/NixOS/nixpkgs/pull/100402#issuecomment-711104497

And this is a people that is very involved with NixOS, so there should be users out there using it without knowing it or care.

TredwellGit replied:
But that case is solved by using `nouveau` or keeping `nvidia` updated.

See: https://www.phoronix.com/scan.php?page=news_item&px=XServer-Abandonware

thiagokokada replied:
> But that case is solved by using nouveau or keeping nvidia updated.

Or not. This is very new GPU and nouveau doesn't seem to support it yet: https://nouveau.freedesktop.org/CodeNames.html

> See: https://www.phoronix.com/scan.php?page=news_item&px=XServer-Abandonware

Yeah, it is not being updated often. However it doesn't mean it will die tomorrow, next years or in the next 5-10 years. X11 will probably continue to live forever, because there is tons of old software and servers out there.

TredwellGit replied:
I disagree with adding `vesa`, but I will add it and submit a pull request to update xorg.xf86videovesa later today.

TredwellGit replied:
Apparently, the latest version of xf86-video-vesa does not even work on UEFI.
https://lists.x.org/archives/xorg-announce/2020-September/003060.html:
>We now refuse to run on machines with UEFI firmware (on Linux only, patches welcome for other OSes) since it won't work in the general case and you probably have a kernel framebuffer driver running already.
>All users are recommended to upgrade, ideally to a better video card and/or driver.

Do you still want me to add it?

TredwellGit replied:
>We also only attempt to use 24bpp if the alternative would be pseudocolor, but note that since xserver 1.20 24bpp will not work at all.

Nixpkgs uses xserver > 1.20.

thiagokokada replied:
> Do you still want me to add it?

I still see value (because in the worst case the user will have no video, but in the slightly better case the user will have video but no accel), but:

> We now refuse to run on machines with UEFI firmware (on Linux only, patches welcome for other OSes) since **it won't work in the general case and you probably have a kernel framebuffer driver running already**.
> All users are recommended to upgrade, ideally to a better video card and/or dri

So maybe we should add `fbdev` as fallback instead? I also found this article about Fedora changing the default: https://www.phoronix.com/scan.php?page=news_item&px=Fedora-VESA-FBDEV-Planning

But I don't know if they're setting `fbdev` or using something else.

TredwellGit replied:
Can you test to make sure that `fbdev` does not supersede `modesetting` on Intel?

TredwellGit replied:
https://github.com/NixOS/nixpkgs/pull/108442 https://github.com/NixOS/nixpkgs/pull/108443

thiagokokada replied:
So I set:

```nix
{
 xserver.videoDrivers = [  "amdgpu" "nouveau" "modesetting" "fbdev" ];
}
```

Did a rebuild, `sudo systemctl restart display-manager.service` and:

```sh
~
❯ cat /var/log/X.0.log | grep fbdev
[ 49210.177] (**) |-->Screen "Screen-fbdev[0]" (3)
[ 49210.177] (**) |   |-->Device "Device-fbdev[0]"
[ 49210.177] (==) No monitor specified for screen "Screen-fbdev[0]".
[ 49210.177] (**) ModulePath set to "/nix/store/d4ggz00ranwmx2isjjx5zmdv0gpnyl7s-xf86-input-libinput-0.28.2/lib/xorg/modules/input,/nix/store/qhgh8v03j4q2hj34ri42jdd1bl2cwll6-xf86-video-amdgpu-19.0.1/lib/xorg/modules/drivers,/nix/store/j1kpqqfqlhywhh0v41d7l6aq4a9cj5b3-xf86-video-nouveau-1.0.15/lib/xorg/modules/drivers,/nix/store/30a13n1yhqpx3fk18dxj7926rm9by65n-xf86-video-fbdev-0.5.0/lib/xorg/modules/drivers,/nix/store/9hzf53n2sd5y4d5ffbkwa8h4axg47b11-xorg-server-1.20.10/lib/xorg/modules,/nix/store/9hzf53n2sd5y4d5ffbkwa8h4axg47b11-xorg-server-1.20.10/lib/xorg/modules/drivers,/nix/store/9hzf53n2sd5y4d5ffbkwa8h4axg47b11-xorg-server-1.20.10/lib/xorg/modules/extensions,/nix/store/fhwz85bwfrv12xqkyjd8jbc3av1imccg-xf86-input-evdev-2.10.6/lib/xorg/modules/input"
[ 49210.185] (II) LoadModule: "fbdev"
[ 49210.185] (II) Loading /nix/store/30a13n1yhqpx3fk18dxj7926rm9by65n-xf86-video-fbdev-0.5.0/lib/xorg/modules/drivers/fbdev_drv.so
[ 49210.185] (II) Module fbdev: vendor="X.Org Foundation"
[ 49210.185] (II) FBDEV: driver for framebuffer: fbdev
[ 49210.197] (WW) Falling back to old probe method for fbdev
[ 49210.197] (II) Loading sub module "fbdevhw"
[ 49210.197] (II) LoadModule: "fbdevhw"
[ 49210.197] (II) Loading /nix/store/9hzf53n2sd5y4d5ffbkwa8h4axg47b11-xorg-server-1.20.10/lib/xorg/modules/libfbdevhw.so
[ 49210.197] (II) Module fbdevhw: vendor="X.Org Foundation"
[ 49210.310] (II) UnloadModule: "fbdev"
[ 49210.310] (II) Unloading fbdev
[ 49210.310] (II) UnloadSubModule: "fbdevhw"
[ 49210.310] (II) Unloading fbdevhw

~
❯ cat /var/log/X.0.log | grep modesetting
[ 49210.177] (**) |-->Screen "Screen-modesetting[0]" (2)
[ 49210.177] (**) |   |-->Device "Device-modesetting[0]"
[ 49210.177] (==) No monitor specified for screen "Screen-modesetting[0]".
[ 49210.185] (II) LoadModule: "modesetting"
[ 49210.185] (II) Loading /nix/store/9hzf53n2sd5y4d5ffbkwa8h4axg47b11-xorg-server-1.20.10/lib/xorg/modules/drivers/modesetting_drv.so
[ 49210.185] (II) Module modesetting: vendor="X.Org Foundation"
[ 49210.185] (II) modesetting: Driver for Modesetting Kernel Drivers: kms
	"Screen-modesetting[0]" for depth/fbbpp 24/32
```

So yeah, it works. Just to make sure I also did:

```nix
{
 xserver.videoDrivers = [  "fbdev" "modesetting" ];
}
```

In this case, the `fbdev` was selected as the default driver. BTW, it was nice because I also guarantee that `fbdev` works as a good fallback: I had full resolution desktop (but my composition didn't work, but I think this was expected).

TredwellGit replied:
This is also good because using `vesa` prevents fixing https://github.com/NixOS/nixpkgs/issues/43984 according to https://wiki.ubuntu.com/X/Rootless. The fact that the order is respected is also nice since it means old AMD GPUs can be supported by default as well.

Atemu commented on nixos/modules/services/x11/xserver.nix:
```diff
@@ -248,11 +248,10 @@ in
         type = types.listOf types.str;
-        # !!! We'd like "nv" here, but it segfaults the X server.
-        default = [ "radeon" "cirrus" "vesa" "modesetting" ];
+        default = [ "amdgpu" "modesetting" ];
```
Wouldn't amdgpu be selected by modesetting automatically? Doesn't hurt to explicitly prioritise it I guess.

TredwellGit replied:
xf86-video-amdgpu is a different DDX than modesetting so how would it be selected without it being installed?

TredwellGit commented:
What needs to be done to get this merged?

ajs124 commented:
This should probably have a changelog entry, besides that, it seems like a sensible change to me.

TredwellGit commented:
You mean release notes entry, right?

ajs124 commented:
Yes, sorry.

TredwellGit commented:
I will write a release notes entry later today.

TredwellGit committed:
nixos/xserver: use modern video drivers

cirrus is obsolete: https://www.vintage3d.org/cirrus.php
nv is obsolete: https://www.phoronix.com/scan.php?page=article&item=nvidia_kills_nv&num=1
vesa is obsolete: https://www.phoronix.com/scan.php?page=news_item&px=Nzc3Nw
ati and ati_unfree are superseded by amdgpu and amdgpu-pro: https://wiki.gentoo.org/wiki/ATI_FAQ#Is_my_AMD.2FATI_board_supported.3F
nouveau and fbdev added for better fallback support.

thiagokokada commented on nixos/doc/manual/release-notes/rl-2103.xml:
```diff
@@ -335,6 +335,11 @@ http://some.json-exporter.host:7979/probe?target=https://example.com/some/json/e
        official documentation</link> of the json_exporter.
      </para>
    </listitem>
+   <listitem>
+     <para>
+       <xref linkend="opt-services.xserver.videoDrivers" /> no longer uses the <literal>cirrus</literal> and <literal>vesa</literal> device dependent X drivers by default.
+     </para>
+   </listitem>
```
I would make this slightly more complete:

>  `opt-services.xserver.videoDriver` no longer sets deprecated `cirrus` and `vesa` drivers, but now enables both `amdgpu` and `nouveau` drivers by default.

thiagokokada commented:
@TredwellGit Are you still interested in merging this PR? Needs rebasing and fixing the merge conflict.

TredwellGit commented:
Yes, just have been busy.

xaverdh commented:
I rebased it: https://github.com/NixOS/nixpkgs/pull/111551

TredwellGit commented:
It is not rebasing it that is the problem. It is being required to write documentation for a proprietary graphics driver that not even the maintainers of that driver want to write.

xaverdh commented:
> It is not rebasing it that is the problem. It is being required to write documentation for a proprietary graphics driver that not even the maintainers of that driver want to write.

I must have missed something, documentation on what needs to be written? Do you refer to `amdgpu-pro`, `nvidia` or something else?

TredwellGit commented:
https://github.com/NixOS/nixpkgs/pull/100402#discussion_r551918037

thiagokokada commented:
> It is not rebasing it that is the problem. It is being required to write documentation for a proprietary graphics driver that not even the maintainers of that driver want to write.

> https://github.com/NixOS/nixpkgs/pull/100402/files/541566b878d50f3669a03f1dcaefe494615e3a3c#diff-d8bd1210dab67b94b4639b9835d4077e1e28ef945212cfa0f20cd5993bfd03b1

Are you saying that about my comments? I am not requiring you do to anything, I don't even have approval/veto powers here (no commit access). That was just a suggestion.

BTW, it is not like I asked you to do document something very obscure either, I just asked for add something similar that already existed for `ati_unfree` for `amdpro`.

TredwellGit commented:
I am going to close in favor of https://github.com/NixOS/nixpkgs/pull/111551. Someone who wants to document `amdgpu-pro` can do so in a different pull request.
